### PR TITLE
Extracted ConnectionInterface from yii\db\Connection yiisoft/yii2#14997

### PR DESCRIPTION
**Pull request by njasm**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #14997

### CODE REVIEW

**Comment by njasm** [comment]
I didn't extracted `yii\db\Connection` `EVENT_*` constants to the interface yet, I think we should, but would like to know your opinion about it.

**Comment by sergeymakinen** [comment]
IMO, events belong to `Component`, first we need to define `ComponentInterface` then it's okay to move the constants.

File: framework/db/ConnectionInterface.php [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+namespace yii\db;
+
+use yii\base\NotSupportedException;
+
+/**
+ * ConnectionInterface defines a connection to a database.
+ *
+ * @author Nelson J Morais <njmorais@gmail.com>
+ * @package yii\db
+ */
```
Comment (sergeymakinen) [inline framework/db/ConnectionInterface.php]: Need `@since` as the interface is new.

Comment (njasm) [reply framework/db/ConnectionInterface.php]: done! used 2.0.15 since that's the version on #14997 issue.

File: framework/db/ConnectionInterface.php [hunk 2]
```diff
@@ -13,6 +13,7 @@
  * ConnectionInterface defines a connection to a database.
  *
  * @author Nelson J Morais <njmorais@gmail.com>
+ * @since 2.0.15
  * @package yii\db
```
Comment (sergeymakinen) [inline framework/db/ConnectionInterface.php]: Yii doesn't use `@package` AFAIK.

**Review by sergeymakinen — APPROVED** [review]

**Review by dmirogin — APPROVED** [review]

File: framework/db/ConnectionInterface.php (hunk 2 above)
Comment (njasm) [reply framework/db/ConnectionInterface.php]: @sergeymakinen you're right, my bad! done @package tag removed.

**Comment by njasm** [comment]
@sergeymakinen  about event's constants: interesting opinion, I surely can understand that events should be "put" on `Component` but then again, take the context of `yii\db\Connection` it does make more sense to me to have constants like: `EVENT_AFTER_OPEN`, `EVENT_BEGIN_TRANSACTION`, `EVENT_ROLLBACK_TRANSACTION` and `EVENT_COMMIT_TRANSACTION` in the `ConnectionInterface` then  `ComponentInterface`, it's just that Component is too generic...

Reading Component class, and what it does/manages, I surely understand your point of view..

**Comment by njasm** [comment]
@sergeymakinen , @dmirogin I've pushed the extraction of Logger interface, can you please review it again? thanks...

**Comment by sergeymakinen** [comment]
@njasm Em, I meant that we need to define a Component interface, so the ConnectionInterface could inherit it and additionally define the constants.

**Comment by njasm** [comment]
> @njasm Em, I meant that we need to define a Component interface, so the ConnectionInterface could inherit it and additionally define the constants.

ah, ok, thanks for the clarification! 👍

File: framework/log/LoggerInterface.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
+<?php
+/**
```
Comment (dmirogin) [inline framework/log/LoggerInterface.php]: phpstorm garbage. I disable this thing in settings always.

**Review by dmirogin — APPROVED** [review]

File: framework/log/LoggerInterface.php (hunk 3 above)
Comment (njasm) [reply framework/log/LoggerInterface.php]: damn it. I should also do it. I've removed that notice and added the yii copyright notice!

File: framework/log/Logger.php [hunk 4]
```diff
@@ -41,7 +41,7 @@
  * @author Qiang Xue <qiang.xue@gmail.com>
  * @since 2.0
  */
-class Logger extends Component
+class Logger extends Component implements LoggerInterface
```
Comment (samdark) [inline framework/log/Logger.php]: Logger doesn't belong to this pull request.

**Review by samdark — COMMENTED** [review]
Overall looks OK.

**Review by samdark — APPROVED** [review]

**Comment by njasm** [comment]
@samdark  @dmirogin removed LoggerInterface commits from this pull request, create #15321 for that matter..

File: framework/db/ConnectionInterface.php [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+ */
+
+namespace yii\db;
+
+use yii\base\NotSupportedException;
+
+/**
+ * ConnectionInterface defines a connection to a database.
+ *
+ * @author Nelson J Morais <njmorais@gmail.com>
```
Comment (SilverFire) [inline framework/db/ConnectionInterface.php]: API was originally authored by Qiang. I think his authoring copyright should not be completely removed.

```
@author Qiang Xue <qiang.xue@gmail.com>
```

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/db/ConnectionInterface.php (hunk 5 above)
Comment (njasm) [reply framework/db/ConnectionInterface.php]: @SilverFire done! I removed my @author tag, can you review it if it's ok this way? thanks.

**Comment by cebe** [comment]
Putting all the different features that the Connection class currently offers into one Interface does not make sense. There may be connection implementations that have no transactions (NoSQL), do not support cache or do not have master/slave splitting. Adding the interface only to make DI work in some way is missuse of it. An interface should define a contract between classes, not serve as a placeholder for DI (at least not as its primary purpose).

**Comment by samdark** [comment]
Therefore, multiple interfaces?
